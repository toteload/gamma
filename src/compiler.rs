use crate::{
    ast::{NodeId, NodeIdGenerator},
    ink_codegen::CodeGenerator,
    parser::Parser,
    semantics::validate_semantics,
    source_location::SourceSpan,
    string_interner::StringInterner,
    type_check::type_check,
    types::{TypeInterner, TypeToken},
};
use std::collections::HashMap;

pub trait PrintableError {
    fn print(&self, context: &Context);
}

pub struct Options {
    pub optimize: bool,
}

pub struct Context {
    pub id_generator: NodeIdGenerator,

    pub symbols: StringInterner,
    pub types: TypeInterner,

    pub ast_spans: HashMap<NodeId, SourceSpan>,
    pub ast_types: HashMap<NodeId, TypeToken>,
}

impl Context {
    pub fn new() -> Self {
        Context {
            id_generator: NodeIdGenerator::new(),
            symbols: StringInterner::new(),
            types: TypeInterner::new(),
            ast_spans: HashMap::new(),
            ast_types: HashMap::new(),
        }
    }

    pub fn compile(
        &mut self,
        source: &str,
        options: &Options,
    ) -> Result<String, Vec<Box<dyn PrintableError>>> {
        let mut parser = Parser::new(
            source,
            &mut self.symbols,
            &mut self.id_generator,
            &mut self.ast_spans,
        );

        let items = parser.parse_items().map_err(|e| {
            let err = Box::new(e) as Box<dyn PrintableError>;
            vec![err]
        })?;

        validate_semantics(&self, &items)?;

        type_check(&items, &mut self.types, &mut self.ast_types).map_err(|errs| {
            errs.into_iter()
                .map(|e| Box::new(e) as Box<dyn PrintableError>)
                .collect::<Vec<Box<dyn PrintableError>>>()
        })?;

        let ctx = inkwell::context::Context::create();
        let mut codegen = CodeGenerator::new(&ctx, &self.symbols, &self.ast_types, &self.types);
        let output = codegen.compile(&items, options.optimize).map_err(|e| {
            let err = Box::new(e) as Box<dyn PrintableError>;
            vec![err]
        })?;

        Ok(output)
    }
}
