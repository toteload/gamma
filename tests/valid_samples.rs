mod valid_samples {

    use gamma::compiler::{Context, Options};
    use insta::assert_snapshot;
    use std::fs;

    #[test]
    fn basic_arithmetic() {
        let contents = fs::read_to_string("tests/valid_samples/basic_arithmetic.gamma").unwrap();

        let mut context = Context::new();
        let result = context.compile(&contents, &Options { optimize: false });

        let Ok(output) = result else {
            let Err(errors) = result else { unreachable!() };

            for e in &errors {
                e.print(&context);
            }

            panic!(
                "Compilation of sample \"basic_arithmetic\" resulted in {} error(s)",
                errors.len()
            );
        };

        assert_snapshot!("basic_arithmetic", output);
    }

    #[test]
    fn basic_main() {
        let contents = fs::read_to_string("tests/valid_samples/basic_main.gamma").unwrap();

        let mut context = Context::new();
        let result = context.compile(&contents, &Options { optimize: false });

        let Ok(output) = result else {
            let Err(errors) = result else { unreachable!() };

            for e in &errors {
                e.print(&context);
            }

            panic!(
                "Compilation of sample \"basic_main\" resulted in {} error(s)",
                errors.len()
            );
        };

        assert_snapshot!("basic_main", output);
    }

    #[test]
    fn sum_loop() {
        let contents = fs::read_to_string("tests/valid_samples/sum_loop.gamma").unwrap();

        let mut context = Context::new();
        let result = context.compile(&contents, &Options { optimize: false });

        let Ok(output) = result else {
            let Err(errors) = result else { unreachable!() };

            for e in &errors {
                e.print(&context);
            }

            panic!(
                "Compilation of sample \"sum_loop\" resulted in {} error(s)",
                errors.len()
            );
        };

        assert_snapshot!("sum_loop", output);
    }
}
