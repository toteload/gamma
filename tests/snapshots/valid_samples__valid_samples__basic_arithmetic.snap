---
source: tests/valid_samples.rs
expression: output
---
; ModuleID = 'main'
source_filename = "main"

define i64 @main() {
entry:
  ret i64 11
}

